### DISCUSSION TITLE
psycopg2 vs psycopg3: difference in column no and table oid in some scenarios

### DISCUSSION THREAD

**Discussion by khushboovashi** [psycopg/psycopg#429] [General]
I found different results in psycpg2 and psycopg3 for table oid and column no. for below queries.

To get the table column, I used cursor.description for psycopg2 and pgresult.ftablecol(col_index) for psycopg3.

Query:  SELECT oid FROM test_table;

Result for table_cloumn:
psycopg3 = -2 
psycopg2 = None

=========================================
To get the table oid, I used cursor.description for psycopg2 and pgresult.ftable(col_index) for psycopg3.

Query: SELECT 1;

Result for table_oid:
Psycopg3 = 0
Psycopg2 = None

So, I just want to confirm that this is intentional.

**Comment by dvarrazzo** [comment]
I would say I can see where these differences are coming from, yes.

Psycopg 3 doesn't expose a processed version of `PQftable()` and `PQftablecol()`. Psycopg2 does some processing on those value, which you can see in `_make_column()` in `psycopg/pqpath.c`.

- In `PQftable()` -> `table_oid`, it converts `InvalidOid` to None.
- In `PQftablecol()` -> `table_column`, it ignores results < 0.
 
The former is reasonable: InvalidOid means pretty much "I don't know". The latter is a shortcoming: maybe I didn't know at time what the cols < 0 were, but they make sense: they are system columns. The info you get from the low-level functions in psycopg 3 is more complete.

```
piro=# select attnum, attname from pg_attribute where attrelid = 'pg_class'::regclass order by attnum limit 10;
┌────────┬──────────────┐
│ attnum │   attname    │
├────────┼──────────────┤
│     -6 │ tableoid     │
│     -5 │ cmax         │
│     -4 │ xmax         │
│     -3 │ cmin         │
│     -2 │ xmin         │
│     -1 │ ctid         │
│      1 │ oid          │
│      2 │ relname      │
│      3 │ relnamespace │
│      4 │ reltype      │
└────────┴──────────────┘
(10 rows)
```

```python
# In psycopg2:
>>> cur.execute("select xmin from pg_class limit 1")
>>> cur.description[0].table_column
None

# In Psycopg 3:
>>> cur.execute("select xmin from pg_class limit 1")
>>> cur.pgresult.ftablecol(0)
-2
```

**Reply by khushboovashi** [reply]
Got it. Thanks.
